rcl_action/include/rcl_action/action_server.h

```diff
@@ -0,0 +1,727 @@
+ * Passing a different type produces undefined behavior and cannot be checked
+ * by this function and therefore no deliberate error will occur.
+ *
+ * If the caller intends to send an 'accepted' response, before calling this function
+ * the caller should use rcl_action_accept_new_goal() to get a rcl_action_goal_handle_t
+ * for future interaction with the goal (e.g. publishing feedback and canceling the goal).
+ *
+ * This function is thread safe so long as access to both the action server and the
+ * `ros_goal_response` are synchronized.
+ * That means that calling rcl_action_send_goal_response() from multiple threads is
+ * allowed, but calling rcl_action_send_goal_response() at the same time as non-thread safe
+ * action server functions is not, e.g. calling rcl_action_send_goal_response() and
+ * rcl_action_server_fini() concurrently is not allowed.
+ * Before calling rcl_action_send_goal_response() the `ros_goal_request` can change and
+ * after calling rcl_action_send_goal_response() the `ros_goal_request` can change, but it
+ * cannot be changed during the rcl_action_send_goal_response() call.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | Yes [1]
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ * <i>[1] for unique pairs of action servers and responses, see above for more</i>
+ *
+ * \param[in] action_server handle to the action server that will make the goal response
+ * \param[in] ros_goal_response a ROS goal response message to send
+ * \return `RCL_RET_OK` if the response was sent successfully, or
+ * \return `RCL_RET_INVALID_ARGUMENT` if any arguments are invalid, or
+ * \return `RCL_RET_ACTION_SERVER_INVALID` if the action server is invalid, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_send_goal_response(
+  const rcl_action_server_t * action_server,
+  const void * ros_goal_response);
+
+/// Accept a new goal using a rcl_action_server_t.
+/**
+ * This is a non-blocking call.
+ *
+ * Creates and returns a pointer to a goal handle for a newly accepted goal.
+ * If a failure occurs, `NULL` is returned and an error message is set.
+
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Yes
+ * Thread-Safe        | Yes
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] action_server the action server that is accepting the goal
+ * \param[in] goal_info a message containing info about the goal being accepted
+ * \param[in] error_msg_allocator a valid allocator or `NULL`
+ * \return a pointer to a new goal handle representing the accepted goal, or
+ * \return `NULL` if a failure occured.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_action_goal_handle_t *
+rcl_action_accept_new_goal(
+  const rcl_action_server_t * action_server,
+  const rcl_action_goal_info_t * goal_info,
```

Comment:
@jacobperron so, if I follow the logic correctly, moments after a successful `rcl_action_take_goal_request()` call, `rcl_action` caller must populate an `rcl_action_goal_info_t` instance appropriately to let the server know of the goal. Is that right? If so, it'd be nice to be explicit about it.

Reply:
Yes, this is correct. I will makes sure this is documented. And add an init/fini for the `rcl_action_goal_info_t`